DigitalLearningSolutions.Web/Controllers/LoginController.cs

```diff
@@ -40,37 +41,82 @@ public IActionResult Index(LoginViewModel model)
                 return View("Index", model);
             }
 
             var (verifiedAdminUser, verifiedDelegateUsers) =
-                loginService.VerifyUsers(model.Password, unverifiedAdminUser, unverifiedDelegateUsers);
-
+                loginService.VerifyUsers(model.Password, adminUser, delegateUsers);
             if (verifiedAdminUser == null && verifiedDelegateUsers.Count == 0)
             {
                 ModelState.AddModelError("Password", "The password you have entered is incorrect.");
                 return View("Index", model);
             }
 
-            var approvedDelegateUser = verifiedDelegateUsers.FirstOrDefault(du => du.Approved);
-
-            if (verifiedAdminUser == null && approvedDelegateUser == null)
+            var approvedDelegateUsers = verifiedDelegateUsers.Where(du => du.Approved).ToList();
+            if (verifiedAdminUser == null && !approvedDelegateUsers.Any())
             {
                 return View("AccountNotApproved");
             }
 
-            verifiedAdminUser ??= loginService.GetVerifiedAdminUserAssociatedWithDelegateUser(verifiedDelegateUsers.First(), model.Password);
-            
-            LogIn(verifiedAdminUser, approvedDelegateUser, model.RememberMe);
-            return RedirectToAction("Index", "Home");
+            verifiedAdminUser ??=
+                loginService.GetVerifiedAdminUserAssociatedWithDelegateUser(verifiedDelegateUsers.First(),
+                    model.Password);
+
+            var availableCentres = GetAvailableCentres(verifiedAdminUser, approvedDelegateUsers);
+            if (availableCentres.Count == 1)
+            {
+                return LogIn(verifiedAdminUser, approvedDelegateUsers.FirstOrDefault(), model.RememberMe);
+            }
+
+            SetTempDataForChooseACentre(model.RememberMe, verifiedAdminUser, approvedDelegateUsers);
+            ChooseACentreViewModel chooseACentreViewModel = new ChooseACentreViewModel
+                { CentreUserDetails = availableCentres };
+            return View("ChooseACentre", chooseACentreViewModel);
+        }
+
+        [HttpGet]
+        public IActionResult ChooseACentre(int centreId)
+        {
+            var rememberMe = (bool)TempData["RememberMe"];
+            var adminAccount = TempData.Get<AdminUser>();
+            var approvedDelegateAccounts = TempData.Get<List<DelegateUser>>();
+            TempData.Clear();
+
+            var adminAccountForChosenCentre = adminAccount?.CentreId == centreId ? adminAccount : null;
+            var delegateAccountForChosenCentre =
+                approvedDelegateAccounts?.FirstOrDefault(du => du.CentreId == centreId);
+
+            return LogIn(adminAccountForChosenCentre, delegateAccountForChosenCentre, rememberMe);
+        }
+
+        private List<CentreUserDetails> GetAvailableCentres(AdminUser? adminUser, List<DelegateUser> delegateUsers)

```

Comment:
I wonder whether this would be better in a helper or a service to make unit testing easier.. 🤔 What do you think?